### DISCUSSION TITLE
Proto working w/ multiple LRAs

### DISCUSSION THREAD

**Discussion by PWPInnovator898** [kriswilk/buzzah#5] [Show and tell]
Attached is main board - now working w/ both 235 Hz and 170 Hz LRAs.
RJ-45s + Ethernet cables connect to gloves.
Next step is to integrate w/ gloves.
![image](https://github.com/kriswilk/buzzah/assets/128418916/0aca3fbd-ec74-4704-9071-074a88f486b2)

**Comment by kriswilk** [comment]
This looks great! I like the use of RJ45 for tethers...I was thinking of the same thing to keep things tidy and easy to disconnect/reconnect. Thanks for sharing!

**Reply by PWPInnovator898** [reply]
Adjustable drive levels and frequency will be important features to calibrate vibration intensity for users and to the baseline targets published by Stanford.

**Reply by kriswilk** [reply]
In case you weren't already aware, amplitude is already adjustable via the AMPLITUDE constant near the top of `code.py`. Range is 0-127 with the default being 127 (full power). Among other quality-of-life improvements, I'll probably change that to a percentage (0-100) like the jitter. I also considered making the amplitude of each motor adjustable individually, in case there is substantial variability between actuators. At least with cheap ERMs, that may be an issue. I think the LRAs are much more consistent.

**Comment by PWPInnovator898** [comment]
Great - thank you! I saw the amplitude constant definition and will adjust
accordingly. I agree with your LRA vs ERMs comments.

One other note - more recent publications reveal "double random
<https://www.researchgate.net/publication/360032158_Long-Lasting_Desynchronization_of_Plastic_Neuronal_Networks_by_Double-Random_Coordinated_Reset_Stimulation>"
(time/jitter _and_ amplitude randomization) stimulation may also improve
efficacy. Of course, the performance of a specific LRA or ERM may not be
capable of responding to these parameter changes simultaneously and may
also be frequency and tactor dependent, but having the dual randomization
capability would expand upon your truly adaptable platform.  My
observations so far are revealing that one existing BT LE-based glove
designs may not be able to support this reliably.   Your thoughts?

I'll continue to test the LRAs I have to let you know how they perform
(i.e., intensity) vs.a pair of the original trial gloves.

On Thu, May 11, 2023 at 8:28 AM Kris Wilk ***@***.***> wrote:

> In case you weren't already aware, amplitude is already adjustable via the
> AMPLITUDE constant near the top of code.py. Range is 0-127 with the
> default being 127 (full power). Among other quality-of-life improvements,
> I'll probably change that to a percentage (0-100) like the jitter. I also
> considered making the amplitude of each motor adjustable individually, in
> case there is substantial variability between actuators. At least with
> cheap ERMs, that may be an issue. I think the LRAs are much more consistent.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kriswilk/buzzah/discussions/5#discussioncomment-5874011>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A6TYIZDNT7HXUXZJWKRFH4LXFTZRRANCNFSM6AAAAAAX5JG7OM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Thanks and best regards,
Phillip Callahan

**Reply by PWPInnovator898** [reply]
Would modifying this register in open loop mode get us the output increase we may need?
![image](https://github.com/kriswilk/buzzah/assets/128418916/8e47cf25-b3fc-495d-ad2c-aa97325a5af1)

**Reply by PWPInnovator898** [reply]
![image](https://github.com/kriswilk/buzzah/assets/128418916/8900b1bf-8d1e-4bda-b90e-4dc6fc20d628)

**Reply by kriswilk** [reply]
Adding the double-random option is trivial, but I want to iron out some other details first. Regarding the datasheet references:

- OD_CLAMP definitely needs to be set correctly to prevent overloading the actuator. However, once set, it's the RTP_INPUT value that should be used to adjust the amplitude.
- DATA_FORMAT_RTP defaults to 0, which simply means that RTP_INPUT is interpreted as a value between +/-127. Zero is no vibration, +/- 127 is maximum vibration. For an ERM, polarity determines the direction of motor rotation. For an LRA it determines the phase. Neither is important for this application so we just use the range 0-127.

Changing DATA_FORMAT_RTP to 1 would just make the range be 0-255, with 255 being the maximum vibration. It wouldn't make it any stronger.

I've done some work on improving the granularity of control over the actuators. I've made a separate "dev" code branch for this work until it's ready to publish. But you can try it by downloading:

https://raw.githubusercontent.com/kriswilk/buzzah/dev/code.py

Changes include:

- allows specification of ERM or LRA via a code constant
- runs both types strictly in OPEN-LOOP for more control (at the expense of auto-braking and overdrive, but I think it's worth it)
- requires specification of desired driving frequency (affects LRA only), defaulting to 235 Hz for the (currently-) suggested LRA
- requires specification of peak voltage, defaulting to about 2.6 V which is the allowed limit for the suggested LRA

The peak voltage was something I hadn't paid attention to... if you look at the datasheet for the specified LRA it's only 1.8 V (RMS) or about 2.6 V peak. The DRV2605 defaults to about 3 V peak, too high to use in open-loop. 

One thing worth noting... I checked the actual output frequency when I set my driver to 250 Hz and found it to be about 260 Hz. I suspect this is due to the driver not having any accurate timebase (eg. crystal oscillator). It's probably using a simple RC oscillator that could be off by a few percent. I may ask TI about that to confirm. But what this means is that if you want precisely 250 Hz you need to tune the frequency setting. I'm still thinking about easy ways to do this...

**Comment by PWPInnovator898** [comment]
Thanks for the notes - very helpful, and I agree w/ you re:  double-random
option - adding it in can come later.
The 170 Hz tactors I have and optionally may use are rated to 2.5 Vrms
(3.54 Vpeak), so I'll be careful and cautious.
Also, thanks for all the embedded code comments, too.  I'll give the new
code a try and let you know how  it goes.

On Fri, May 12, 2023 at 11:24 AM Kris Wilk ***@***.***> wrote:

> Adding the double-random option is trivial, but I want to iron out some
> other details first. Regarding the datasheet references:
>
>    - OD_CLAMP definitely needs to be set correctly to prevent overloading
>    the actuator. However, once set, it's the RTP_INPUT value that should be
>    used to adjust the amplitude.
>    - DATA_FORMAT_RTP defaults to 0, which simply means that RTP_INPUT it
>    is interpreted as a value between +/-127. Zero is no vibration, +/- 127 is
>    maximum vibration. For an ERM this has the additional effect of spinning
>    the motor in either direction.
>
> Changing DATA_FORMAT_RTP to 1 would just make the range be 0-255, with 255
> being the maximum vibration. It wouldn't make it any stronger.
>
> I've done some work on improving the granularity of control over the
> actuators. I've made a separate "dev" code branch for this work until it's
> ready to publish. But you can try it by downloading:
>
> https://raw.githubusercontent.com/kriswilk/buzzah/dev/code.py
>
> Changes include:
>
>    - allows specification of ERM or LRA via a code constant
>    - runs both types strictly in OPEN-LOOP for more control (at the
>    expense of auto-braking and overdrive, but I think it's worth it)
>    - requires specification of desired driving frequency (affects LRA
>    only), defaulting to 235 Hz for the (currently-) suggested LRA
>    - requires specification of peak voltage, defaulting to about 2.6 V
>    which is the allowed limit for the suggested LRA
>
> The peak voltage was something I hadn't paid attention to... if you look
> at the datasheet for the specified LRA it's only 1.8 V (RMS) or about 2.6 V
> peak. The DRV2605 defaults to about 3 V peak, too high to use in open-loop.
>
> One thing worth noting... I checked the actual output frequency when I set
> my driver to 250 Hz and found it to be about 260 Hz. I suspect this is due
> to the driver not having any accurate timebase (eg. crystal oscillator).
> It's probably using a simple RC oscillator that could be off by a few
> percent. I may ask TI about that to confirm. But what this means is that if
> you want precisely 250 Hz you need to tune the frequency setting. I'm still
> thinking about easy ways to do this...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kriswilk/buzzah/discussions/5#discussioncomment-5887427>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A6TYIZGYRSIBPS7YEUTTY7LXFZW5RANCNFSM6AAAAAAX5JG7OM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Thanks and best regards,
Phillip Callahan

**Comment by PWPInnovator898** [comment]
My initial tests of your dev code appear to work perfectly.   For both 235 and 170 Hz LRAs.  I’ll test ERMs and let you know the results. Excellent work and thanks again for your contributions to this community!

**Reply by kriswilk** [reply]
Glad to hear it's working. I've ordered a couple of the 170 Hz LRAs to get a feel for the difference in performance myself.

**Comment by GrayGhost-su** [comment]
Thank you both for the excellent work back and forth to come up with different iterations of this project. Being a part of the Parkinson community I really appreciate your testing and improving this project. My parts are on order and I am looking forward to doing some testing of my own.

**Reply by kriswilk** [reply]
Great, let us know how it goes. After getting the published version working, you may want to try the "dev" version of the code as outlined above...it's got a number of adjustable options that might be useful to you. In the near future, I'll push the code into the main branch for general use.

**Comment by PWPInnovator898** [comment]
I’m back after some time away from my office. Questions:1) can you confirm the recommended procedures to run in closed loop mode and open loop mode?  Your code is well commented but I’d like to make sure I don’t miss anything. 2) can the two modes (open and closed) both support random amplitude stimulation, simultaneously to the random timing (jitter) mode already supported? Thanks again!On May 15, 2023, at 07:07, Kris Wilk ***@***.***> wrote:﻿
Glad to hear it's working. I've ordered a couple of the 170 Hz LRAs to get a feel for the difference in performance myself.

**Reply by PWPInnovator898** [reply]
Hello;
In open loop mode, the peak output voltage is unexpectedly being limited (clamped) to about 1.7Vrms, whereas it should be capable of higher levels (e.g, 2.6Vrms).  
The current dev code is shown below.     
# set peak voltage
    driver._write_u8(0x17, int(ACTUATOR_VOLTAGE / 0.02122))
TI datasheet has: Vpeak = 0.02122 × OD_CLAMP[7:0]
I'm not sure what the issue might be, but I cannot set the max output to be above about 1.7Vrms.
Note: All other open loop tests are working as expected,
Thanks.

**Reply by kriswilk** [reply]
Sorry, I've been away for a while, too.

1) The dev branch firmware doesn't have an option for closed-loop right now. If/when I add that back, doing it properly will need to include the auto-calibration routine specified in the datasheet to optimize the performance of each actuator.

2) I just updated the dev branch with parameters for randomized amplitude, in the form of AMPLITUDE_MIN and AMPLITUDE_MAX. Both default to 100% (fixed amplitude). Give it a try. I've noticed that (at least to me) the perceived output doesn't feel proportional to the amplitude. For instance, at lower amplitudes, the output feels like it drops off non-linearly. BTW, if closed-loop is re-implemented, these settings should work just as well with it.

3) Regarding the output voltage, I'll have to test it again in more detail. How are you measuring the voltage? Multimeter (if so, what type), oscilloscope? I scoped the output a week or two ago when I was adjusting the code and it looked like (at least in open-loop) the driving signal was roughly square. It seemed to be able to reach a peak of up to 3.3 V (which it the input power supply voltage). But to be honest, I was more concerned with the output frequency at the time, so I'll need to recheck...

When I have a chance to test again, I'll comment back here and make corrections to the code if I've flubbed something... 🤣

**Reply by PWPInnovator898** [reply]
Thanks for your detailed replies.  My comments per your list are below:

1) I agree w/ you re: auto-cal inclusion in closed loop mode. The original code seemed to work very well in closed loop mode.  For simplicity, I would be ok with having two code releases, if needed, to avoid the register setting complications needed for proper open loop control.

2) I'll test the new randomized amplitude mode and let you know how it works.

3) I'm currently measuring the output amplitude using a "true rms" AC multimeter (until I can get access to a scope to confirm the multimeter's readings). I also expect a square wave from the haptic driver IC in RTP mode (and it certainly "sounds" like one, too), in contrast to the pure sine wave from the PC audio / class D system (I was using previously), which allows the output swing to about 2.7Vrms (as expected).  
I measured output swing again in open loop using the latest code, yet the output appears to be still clamped around 1.7Vrms.  The good news is the lower output swing in open loop mode appears to be the only remaining operational difference in  the observed vibration intensity differences between my original PC audio-based stimulation system (AND the trial gloves for that matter) and the system I duplicated based on your h/w and s/w.  Once that issue is understood, I'm confident the system you designed will provide all the necessary vibratory performance and flexibility.
Thanks again for your excellent contributions to the community.

**Reply by PWPInnovator898** [reply]
I reviewed the max amplitude function of the RMS multimeter and discovered the readings were misleading.  The amplitude settings in the code appear to be driving to the correct values in open loop mode. I will also try to confirm this with an oscilloscope. What’s left for me is to integrate the LRAs in the glove assembly which I’m still working on.   Vibration intensity range appears to be adequate as compared to the trial gloves.

**Reply by kriswilk** [reply]
That's good to hear. In any case I'll be double-checking when I've got a chance.

**Comment by PWPInnovator898** [comment]
Latest pics of proto system are attached.
![image](https://github.com/kriswilk/buzzah/assets/128418916/b542348d-ec0e-464e-8732-465218888c6a)
Next steps:
- Modifications to run from the 5V USB source from the  RP2040 PCB supply, to eliminate 3.3V supply droop are a WIP, but the initial test worked
- Better ways to simplify and reduce the PCB area compared to this initial revision
- Finish assembling the tactors w/ the gloves
- For portability, the PCB and battery will be enclosed in a protective box with a soft neoprene fabric to form a "VCR fanny pack" with the short ethernet patch cables wrapping around the front of the body to the RJ-45 ports connecting to the gloves
Feedback welcomed.

**Reply by GrayGhost-su** [reply]
Very nice. I really like the RJ45 connector idea.. How are you mounting the LRA's into your finger cots and where did you find the RJ45 connectors for the gloves ? All my parts are finally here, time to start putting together.
Thanks to both for the excellent work on this.

**Reply by kriswilk** [reply]
I agree, very nice job. @PWPInnovator898 can let us know where he got his connectors, but if you search for "rj45 breakout screw terminal" you will find many possible solutions like it. For instance, something like this one is quite nice and easy to rewire if/when leads  break, or to try alternative actuators:

https://www.amazon.ca/Poyiccot-Compatible-Terminal-Connector-Ethernet/dp/B07WKKVZRF

**Reply by PWPInnovator898** [reply]
I used a terminal block-to-rj-45 adapter very similar to the one Mentioned above. Mounting the LRA in the holders is quite a challenge. The ones you see in the picture are handmade in foam to provide the best vibrational energy coupling to the fingertip, which is critical. The next revision will be 3-D printed in foam to make the assembly, more consistent for long-term use.   I will send more information on the design in my next message.  By the way, if you opt for using square 10 mm tactors as I use now, they provide the highest vibrational energy range for the devices I’ve been testing.    I recommend connecting the 5V supply from the RP2040 USB connector directly to the TI haptic driver IC modules.   This avoids the 3.3 V supply droop when operated with these tactors.   The tactor’s are Vybronics’ VLV1040A series.  Digi-Key has them on their website.

**Reply by GrayGhost-su** [reply]
Thank you both for the info, I will check them both out.

**Reply by PWPInnovator898** [reply]
Latest tests w/ 5V USB supply and [Vybronics' VLV1040A LRAs] (https://www.vybronics.com/linear-lra-vibration-motors/v-lv101040a) were positive - very close IMHO to trial VCR gloves' stimulation sensation.  Running system in open loop mode at 2.5Vrms and 235 Hz.  
Next PCB rev will be more compact for better portability.

**Reply by GrayGhost-su** [reply]
PWP how are you connecting to the Vybronics' VLV1040A LRAs pads ? soldering wires to the pad or using a pin connector ? also are you running 5V supply wire to each of the 8 haptic driver boards ?

Thanks GG

**Reply by kriswilk** [reply]
> are you running 5V supply wire to each of the 8 haptic driver boards

I'm guessing that @PWPInnovator898 wired the "5V" pad from the microcontroller board to the "V+" pad on the multiplexer. Then you'd switch the multiplexer's tiny switch to the "V+" setting in order to drive all of its output ports with a 5V power level. You shouldn't need separate wires going to each haptic board.

That said, if the RMS voltage is 2.5 V or less, the standard 3.3V supply should be adequate. If it isn't, maybe the onboard regulator on the microcontroller isn't beefy enough. But I'd be surprised if that were the case.

**Reply by PWPInnovator898** [reply]
GreyGhost: Re: Vybronics' VLV1040A LRAs pads, I've had wires soldered to the pads directly w/ 28/32 AWG.  Painstaking process ...
kriswilk- yes, that's what I "should" have done. And I agree, connecting 5V from uC/USB to the mux should work fine. I went and made 5V connections to each haptic board in case the MCU's on-board regulator begins limiting at higher outputs. But, I'll connect your way on future builds.   Thanks for the helpful suggestion.
Attached are pics of two complete systems. 

1. First one is my larger PCB build (but w/out an enclosure, which is WIP), including battery.  I'm beginning my VCR sessions w/ this version this week.
2. 2nd is a smaller build (w/ identical electronics) from a former colleague, with a much smaller enclosure, and same battery.

S/W continues to work great (thanks Kris!). I'm slowly validating and adding some code to help, e.g., log therapy time and stimulation settings (WIP).   

i am hoping for some positive initial effects but recognizing the ""multi-month" patience  [Dr. Tass continues to mention in his seminars.](https://www.youtube.com/watch?v=XRUOjpm0hKo)

![image](https://github.com/kriswilk/buzzah/assets/128418916/420f0618-e449-49f1-b842-fef9c0f6a6ac)
![image](https://github.com/kriswilk/buzzah/assets/128418916/04ffe0b6-37e5-43fa-8531-b5eaab4e277b)

**Reply by kriswilk** [reply]
Great progress, thanks for sharing. I've just merged the code from the dev branch into the main branch so it's the "official" version.

**Reply by PWPInnovator898** [reply]
Just to confirm, your latest code is here: https://github.com/kriswilk/buzzah/blob/main/code.py ?
I was using a slightly older revision...  
BTW - another PwP friend appears to be getting encouraging therapeutic results w/ the latest system build (the small box shown in the photo above).

**Reply by kriswilk** [reply]
Yes, that link should always be to the latest main branch release.

Glad to hear that there have been positive results!

**Reply by PWPInnovator898** [reply]
Greetings Kris;
Good news --  I continue to hear encouraging therapeutic results from a close PwP friend using the Buzzah h/w!
I was wondering if you could point me to a datalogging tutorial example for the QTP2040 mcu h/w that buzzah uses? Given the encouraging results so far, it would be great to log various session therapy settings and info automatically for use in documenting and optimizing therapy.   My circuitpython experience is growing, but I've yet to find a straightforward example that utilizes the h/w resources now in place. Thanks kindly in advance.

**Reply by kriswilk** [reply]
Sorry for the delay, family summer road trip time...

The QT Py 2040, like all circuitpython boards, has flash storage onboard. In the case of this board, it's 8 MB...plenty to store some text logs. Any files written there are also directly accessible via USB just like the code.py file, so no fancy transfer acrobatics are required. So that's the mechanism I'd use.

The only speedbump is that to write to the flash the code needs exclusive access to it (the PC and the code can't have write access simultaneously). But there's a way to deal with this.

The storage system is outlined with a data logging example here:

https://learn.adafruit.com/circuitpython-essentials/circuitpython-storage

The example has references to particular boards/pins and doesn't specifically include QT Py 2040. Basically you just need to pick an available I/O pin to use as an input to signal who gets write access.

I can help more with this once I'm home later in the week but that ought to get you started.

Kris

**Comment by GrayGhost-su** [comment]
Got the firmware and program with libraries loaded on the uController and everything lights up and each of the motor drivers work. I only have one LRA motor put together but I used that one to check each driver board and all 8 work............ hurray !!!
next up mounting for a smaller form factor and ordering some RJ45 terminal block adapters for cabling to gloves.

Thanks again to both of you ..............
![IMG_0154](https://github.com/kriswilk/buzzah/assets/134098896/def74b43-0967-45d2-a378-165e4a89be27)

**Reply by RonCurePD** [reply]
Hi, a picture is worth a thousand words. I just completed building a unit just like yours (I hope). However, I cannot tell if the LRA motor on my unit is working or not. When the 5V power pack battery is connected, are the LRA motors suppose to make noise, jump about, buzz, or what? My LRA motors seem not active. Am I missing something?
Ron

**Reply by kriswilk** [reply]
Hi @RonCurePD,

If you've assembled the same parts as pictured and are powering the device from a decent USB power source, any connected actuators should vibrate in random sequences. If that's not happening, it's most likely that something about the loading of the code into the microcontroller went wrong. Specifically, [this part](https://github.com/kriswilk/buzzah?tab=readme-ov-file#setup) of the instructions. If in doubt, try following that part again, copying a fresh set of all files to the QT Py board.

It's also worth noting that if you use an actuator different from the one specified, you may need to adjust lines 14, 16, and 18 in `code.py` to match the specifications of the particular device.

**Comment by PWPInnovator898** [comment]
Thank you very much for the tip, Kris.  I will get started on this and hopefully get something figured out by the time you return.  No hurry to respond as you’re on vacation.  :-)Thanks and best regards, PhillipOn Jul 23, 2023, at 07:03, Kris Wilk ***@***.***> wrote:﻿
Sorry for the delay, family summer road trip time...
The QT Py 2040, like all circuitpython boards, has flash storage onboard. In the case of this board, it's 8 MB...plenty to store some text logs. Any files written there are also directly accessible via USB just like the code.py file, so no fancy transfer acrobatics are required. So that's the mechanism I'd use.
The only speedbump is that to write to the flash the code needs exclusive access to it (the PC and the code can't have write access simultaneously). But there's a way to deal with this.
The storage system is outlined with a data logging example here:
https://learn.adafruit.com/circuitpython-essentials/circuitpython-storage
The example has references to particular boards/pins and doesn't specifically include QT Py 2040. Basically you just need to pick an available I/O pin to use as an input to signal who gets write access.
I can help more with this once I'm home later in the week but that ought to get you started.
Kris

**Comment by Neurosmith24** [comment]
@PWPInnovator898 did you end up soldering the breakaway right-angle header pins to the drivers like @kriswilk did or did you just solder wires that go to the ethernet plug and tactors directly onto the DRV2605L Haptic Motor Controllers? (hard to tell from the photos) Thanks!

**Reply by PWPInnovator898** [reply]
Yes I used right angle pins to allow easy replacement

**Comment by RonCurePD** [comment]
Yes indeed, we found out that the problem were the four .mpy files included in the lip folder per https://github.com/kriswilk/buzzah were incompatible. We downloaded the new files from https://circuitpython.org/libraries  and replaced the four .mpy files in the lib folder and the problem was resolved.
Thank you for your help.
Ron Lim
***@***.***
 

    On Monday, June 17, 2024 at 04:46:47 PM PDT, Kris Wilk ***@***.***> wrote:   

 


Hi @RonCurePD,

If you've assembled the same parts as pictured and are powering the device from a decent USB power source, any connected actuators should vibrate in random sequences. If that's not happening, it's most likely that something about the loading of the code into the microcontroller went wrong. Specifically, this part of the instructions. If in doubt, try following that part again, copying a fresh set of all files to the QT Py board.

It's also worth noting that if you use an actuator different from the one specified, you may need to adjust lines 14, 16, and 18 in code.py to match the specifications of the particular device.

**Reply by kriswilk** [reply]
Glad you got it figured out. Last night another user was having trouble and I, too, discovered that the lib files were outdated. I've updated the files in this repo to the correct versions for the latest CircuitPython (9.0.5).

**Comment by PWPInnovator898** [comment]
Hello Kris;

It's been awhile since we last communicated re: Buzzah.
I hope all is well with you.

FYI: In collaboration with TactileDesign <https://github.com/TactileDesign>,
we just posted a new "BlueBuzzah
<https://github.com/PWPInnovator898/BlueBuzzah-Gloves>" repository based in
part on your Buzzah project.

Although this is my first public repository contribution to GitHub, I trust
the references in GitHub properly recognize and attribute your excellent
Buzzah work.  Your Circuit Python code has been quite helpful in adding
Bluetooth to the prototypes.

If not, please let us know.

-- 
Thanks and best regards,
Phillip Callahan
